core/src/main/java/io/ultrabrew/metrics/data/ConcurrentMonoidLongTable.java

```diff
@@ -294,11 +296,30 @@ long index(final String[] tags, final boolean isReading) {
                 // grow tag set
                 synchronized (this) {
                   if (tagIndex >= tagSets.length) {
+                    boolean isEqual = false;
+                    if(tagIndex == tagSets.length) {
+                      isEqual = true;
+                    }
                     final int oldLength = tagSets.length;
                     final int newLength =
                         oldLength > TAGSETS_MAX_INCREMENT ? oldLength + TAGSETS_MAX_INCREMENT
                             : oldLength * 2;
-                    tagSets = Arrays.copyOf(tagSets, newLength);
+                    String[][] newTagSet = new String[newLength][];
+                    int cnt = 0;
+                    for (int i = 0; i < tagSets.length; i++) {
+                      // Skip the entries that would have been freed up (stale)
+                      // System.arraycopy would be faster, but can't use it here due to 
+                      // the conditional inclusion 
+                      if ((tagSets[i].length > 0 && tagSets[i][0] != null)
```

Comment:
Assign `tagSets[i]` to a local variable maybe?

Reply:
Done